src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.cs

```diff
@@ -189,7 +189,16 @@ public Utf8JsonReader(ReadOnlySpan<byte> jsonData, bool isFinalBlock, JsonReader
         /// </exception>
         public bool Read()
         {
-            return _isSingleSegment ? ReadSingleSegment() : ReadMultiSegment();
+            bool retVal = _isSingleSegment ? ReadSingleSegment() : ReadMultiSegment();
+
+            if (!retVal)
```

Comment:
Though this is correct, I am concerned about the performance regression for the overwhelmingly common case. I believe adding a check here is inefficient and we can probably add the checks in more targeted places (for example, we can create a custom `HasMoreDataThrowIfEmpty()` that adds such a check and we can call that inside `ReadSingleSegment()` as the first call instead, along with after skipping the white space).

Let me know if you are up for it, and I can try to help or provide guidance. Otherwise, this fix is fine as a temporary measure.

Reply:
**As first steps, look at this method:** https://github.com/dotnet/corefx/blob/1fd3717d73da3ae7440841e8fded4adbba5b0770/src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.cs#L249-L344

The cases where we have empty JSON can be captures within `HasMoreData`, here:
https://github.com/dotnet/corefx/blob/1fd3717d73da3ae7440841e8fded4adbba5b0770/src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.cs#L253
and here:
https://github.com/dotnet/corefx/blob/1fd3717d73da3ae7440841e8fded4adbba5b0770/src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.cs#L267

We should add something like `HasMoreDataThrowIfEmpty()` and update the two callsites above with this new method (which has the check to throw if there is no more tokens/data instead of returning false).

**The next step is dealing with comments which is a bit trickier:**
Look into ReadFirstToken (particularly ConsumeValue):
https://github.com/dotnet/corefx/blob/1fd3717d73da3ae7440841e8fded4adbba5b0770/src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.cs#L411-L475

Essentially, ConsumeValue is called to read any JSON value. However, if the first one is a comment (and skip comment is enabled), and there is no more data after, we need to throw. One way to do this is to create a ConsumeFirstValue method which is essentially identical to ConsumeValue but has this throw logic. That may not be the best solution, but it should work.

Similarly, you would have to update the ReadOnlySequence code paths (which I will let you investigate), i.e.:
https://github.com/dotnet/corefx/blob/9a0e15d1777de9c951cec081699be03304b99bda/src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs#L79-L177

Feel free to ask questions if you get stuck. Thanks.

Reply:
Thanks. Could you please advice, should it be something like ?
```csharp
private bool HasMoreDataThrowIfEmpty()
{
	bool retVal = HasMoreData();

	if (!retVal && _isFinalBlock && TokenType == JsonTokenType.None)
	{
		ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedJsonTokens);
	}

	return retVal;
}
```

Reply:
I'm not sure that I know how to implement the ConsumeFirstValue 😕

Reply:
> I'm not sure that I know how to implement the ConsumeFirstValue

No problem. I think we can merge this change as is for now and evaluate the performance regression separately. It might be too involved. Thanks for fixing the issue :)